timemachine/fe/single_topology_v3.py

```diff
@@ -0,0 +1,539 @@
+        Force constants and equilibrium angles
+
+    improper_idxs: list of 4-tuples
+        Trefoil idxs for improper torsions
+
+    improper_params: list of 3-tuples
+        Force constant, phase, periods
+
+    dummy_group: set or list of int
+        Atoms to be decoupled
+
+    root_anchor_atom: int
+        A core atom we want to anchor our dummy_group to
+
+    nbor_anchor_atom: int
+        Another core atom connected to root_anchor_atom to build an angle restraint off of.
+
+    Returns
+    -------
+    (bonded_idxs, bonded_params)
+        Returns bonds, angles, and improper idxs and parameters.
+    """
+
+    # copy interactions that involve only root_anchor_atom
+    dummy_bond_idxs = []
+    dummy_bond_params = []
+    dummy_angle_idxs = []
+    dummy_angle_params = []
+    dummy_improper_idxs = []
+    dummy_improper_params = []
+
+    # dummy_group may be a set in certain cases, so sanity check.
+
+    assert len(dummy_group) == len(list(dummy_group))
+    dummy_group = list(dummy_group)
+    # dummy group and anchor
+
+    dga = dummy_group + [root_anchor_atom]
+    for idxs, params in zip(bond_idxs, bond_params):
+        if all([a in dga for a in idxs]):
+            dummy_bond_idxs.append(tuple([int(x) for x in idxs]))  # tuples are hashable etc.
+            dummy_bond_params.append(params)
+    for idxs, params in zip(angle_idxs, angle_params):
+        if all([a in dga for a in idxs]):
+            dummy_angle_idxs.append(tuple([int(x) for x in idxs]))
+            dummy_angle_params.append(params)
+    for idxs, params in zip(improper_idxs, improper_params):
+        if all([a in dga for a in idxs]):
+            dummy_improper_idxs.append(tuple([int(x) for x in idxs]))
+            dummy_improper_params.append(params)
+
+    # (ytz): copy interactions that involve nbor_anchor_atom, if not None
+    # this may be set to None
+    if nbor_anchor_atom is not None:
+        found = False
+        for idxs, params in zip(angle_idxs, angle_params):
+            i, j, k = idxs
+            if (i in dummy_group and j == root_anchor_atom and k == nbor_anchor_atom) or (
+                k in dummy_group and j == root_anchor_atom and i == nbor_anchor_atom
+            ):
+                dummy_angle_idxs.append(tuple([int(x) for x in idxs]))
+                dummy_angle_params.append(params)
+                found = True
+
+        if not found:
+            # User provided a bad nbor_anchor_atom
+            raise MissingAngleError("Missing angle interaction in mol_b")
+
+    return (dummy_bond_idxs, dummy_angle_idxs, dummy_improper_idxs), (
+        dummy_bond_params,
+        dummy_angle_params,
+        dummy_improper_params,
+    )
+
+
+def setup_end_state(ff, mol_a, mol_b, core, a_to_c, b_to_c):
+    """
+    Setup end-state for mol_a with dummy atoms of mol_b attached. The mapped indices will correspond
+    to the alchemical molecule with dummy atoms.
+
+    Parameters
+    ----------
```

Comment:
Nit: docstring -- missing parameter `ff`

Reply:
pausing review here - will resume later tonight (note to self mostly)

Reply:
fixed